pkg/instrument/dqlite.go

```diff
@@ -0,0 +1,125 @@
+package instrument
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"os"
+	"sync"
+	"sync/atomic"
+)
+
+type Listener struct {
+	Network, Address        string
+	AcceptedChan            chan net.Conn
+	netListener             net.Listener
+	bytesWritten, bytesRead atomic.Int64
+	wg                      sync.WaitGroup
+	err                     error
+}
+
+type ListenerMetrics struct {
+	BytesWritten int64
+	BytesRead    int64
+}
+
+func NewListener(network, address string) *Listener {
+	return &Listener{
+		Network:      network,
+		Address:      address,
+		AcceptedChan: make(chan net.Conn),
+		netListener:  nil,
+	}
+}
+
+func (l *Listener) Listen(ctx context.Context) error {
+	if l.netListener != nil {
+		return fmt.Errorf("listener already running")
+	}
+	var lc net.ListenConfig
+	netListener, err := lc.Listen(ctx, l.Network, l.Address)
+	if err != nil {
+		return err
+	}
+	l.netListener = netListener
+
+	l.wg.Add(1)
+	go func() {
+		defer l.wg.Done()
+		for {
+			conn, err := l.netListener.Accept()
+			if ctx.Err() != nil {
+				break
+			}
+			if err != nil {
+				l.err = errors.Join(l.err, err)
+				break
+			}
+			l.AcceptedChan <- conn
+		}
+	}()
+	return nil
+}
+
+func (l *Listener) Connect(ctx context.Context, _ string) (net.Conn, error) {
+	dialer := &net.Dialer{}
+	conn, err := dialer.DialContext(ctx, l.Network, l.Address)
+	if err != nil {
+		return nil, err
+	}
+	if conn, ok := conn.(fileConn); ok {
+		return &metredConn{conn, l}, nil
+	}
+	return nil, fmt.Errorf("can't setup probes")
```

Comment:
I think usually we avoid contractions in error messages so we should `s/can't/cannot/` here

Reply:
It's about the 100th time you correct me there lol